homeassistant/components/greencell/__init__.py

```diff
@@ -0,0 +1,194 @@
+        }
+        if device_id in known:
+            _LOGGER.debug("Device %s already configured", device_id)
+            return
+
+        topic = f"/greencell/evse/{device_id}/cmd"
+        pub_payload = json.dumps({"name": "QUERY"})
+
+        async def _async_send_query() -> None:
+            try:
+                await mqtt.async_wait_for_mqtt_client(hass)
+                await mqtt.async_publish(hass, topic, pub_payload, qos=0, retain=False)
+                _LOGGER.info("Sent QUERY to new device %s", device_id)
+            except HomeAssistantError as err:
+                _LOGGER.error("Error publishing QUERY for %s: %s", device_id, err)
+
+        hass.async_create_task(_async_send_query())
+
+    async def _async_subscribe() -> None:
+        nonlocal unsub
+        try:
+            unsub = await async_subscribe(hass, GREENCELL_DISC_TOPIC, _handle_hello)
+            _LOGGER.debug("Subscribed to discovery topic %s", GREENCELL_DISC_TOPIC)
+        except HomeAssistantError as err:
+            _LOGGER.error("Failed to subscribe to discovery topic: %s", err)
+
+    hass.async_create_task(_async_subscribe())
+
+    def _unsubscribe() -> None:
+        if unsub:
+            unsub()
+
+    return _unsubscribe
+
+
+def wait_for_device_ready(
+    hass: HomeAssistant, serial: str, timeout: float
+) -> tuple[Callable[[], None], asyncio.Event]:
+    """Subscribe to GREENCELL_DISC_TOPIC and device voltage topic.
+
+    Return (unsubscribe_all, event) where event is set on first message.
+    """
+    event = asyncio.Event()
+    unsub_disc: Callable[[], None] | None = None
+    unsub_volt: Callable[[], None] | None = None
+
+    @callback
+    def _on_message(message: Any) -> None:
+        if not event.is_set():
+            event.set()
+            _LOGGER.debug("Received initial message for device %s", serial)
+
+    async def _async_subscribe_all() -> None:
+        nonlocal unsub_disc, unsub_volt
+        try:
+            unsub_disc = await async_subscribe(hass, GREENCELL_DISC_TOPIC, _on_message)
+        except HomeAssistantError:
+            _LOGGER.error("Cannot subscribe to discovery topic for readiness check")
+        try:
+            topic = f"/greencell/evse/{serial}/voltage"
+            unsub_volt = await async_subscribe(hass, topic, _on_message)
+        except HomeAssistantError:
+            _LOGGER.error(
+                "Cannot subscribe to voltage topic %s for readiness check", topic
+            )
+
+    hass.async_create_task(_async_subscribe_all())
+
+    def _unsubscribe_all() -> None:
+        if unsub_disc:
+            unsub_disc()
+        if unsub_volt:
+            unsub_volt()
+
+    return _unsubscribe_all, event
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Setup by YAML configuration is not supported."""
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
```

Comment:
You should call `mqtt.async_wait_for_mqtt_client(hass)` as done in other integrations using the built-in MQTT client.